src/ngraph/pass/constant_folding_non_zero.cpp

```diff
@@ -0,0 +1,235 @@
+        }
+
+        template <typename T>
+        void find_indices(const T* values)
+        {
+            m_current_index = Shape(m_input_shape.size(), 0UL);
+            const auto values_count = shape_size(m_input_shape);
+
+            const T zero_value = T{0};
+            for (size_t i = 0; i + 1 < values_count; ++i)
+            {
+                if (values[i] != zero_value)
+                {
+                    add_to_results(m_current_index);
+                }
+
+                next_index();
+            }
+
+            // check the last element in the input values
+            if (values_count != 0 && values[values_count - 1] != zero_value)
+            {
+                add_to_results(m_current_index);
+            }
+        }
+
+        void generate_all_indices()
+        {
+            m_current_index = Shape(m_input_shape.size(), 0UL);
+            size_t i = 0;
+            const auto values_count = shape_size(m_input_shape);
+            while (i + 1 < values_count)
+            {
+                add_to_results(m_current_index);
+                next_index();
+                ++i;
+            }
+            add_to_results(m_current_index);
+        }
+
+        const Results_t& get_indices() const { return m_results; }
+    private:
+        /// \brief Adds single dimensions of an index into the matching element of the results
+        void add_to_results(const Shape& index)
+        {
+            for (size_t dim = 0; dim < index.size(); ++dim)
+            {
+                m_results.at(dim).push_back(index[dim]);
+            }
+        }
+
+        // Generates an index pointing to the next element in the flattened tensor
+        // It behaves similar to flipping bits when incrementing a binary number
+        inline void next_index()
+        {
+            for (size_t dim = m_current_index.size() - 1; dim >= 0; --dim)
+            {
+                auto& dim_value = m_current_index.at(dim);
+                if (dim_value + 1 == m_input_shape[dim])
+                {
+                    dim_value = 0;
+                }
+                else
+                {
+                    ++dim_value;
+                    return;
+                }
+            }
+        }
+
+    private:
+        const Shape m_input_shape;
+        Results_t m_results;
+        Shape m_current_index;
+    };
+}
+
+template <typename T>
+static shared_ptr<op::Constant> fold_constant_non_zero(const shared_ptr<op::Constant>& data)
+{
+    const auto input_shape = data->get_shape();
+    const auto* input_values = data->get_data_ptr<T>();
+    const bool identical_elems_in_data = data->get_all_data_elements_bitwise_identical();
+
+    if (identical_elems_in_data)
+    {
+        NGRAPH_CHECK(input_values[0] != T{0},
+                     "It's not possible to constant fold a NonZero op with an input containing "
+                     "only zeros.");
```

Comment:
You can't make not being able to fold a constant an error -- if you can't fold, you just don't fold.

Reply:
Done with a test to make sure that CF doesn't happen in this case.